packages/actor-rdf-resolve-hypermedia-links-queue-wrapper-info-occupancy/lib/LinkQueueLogger.ts

```diff
@@ -0,0 +1,209 @@
+        reachability_criteria_dynamic_info: reachabilityCriteriaParentLink === null ?
+          undefined :
+          reachabilityCriteriaParentLink[1],
+      };
+
+      const reachabilityCriteriaCurrentLink = LinkQueueLogger.getReachability(link);
+
+      const statisticLink: IURLStatistic = {
+        url: link.url,
+        reachability_criteria: reachabilityCriteriaCurrentLink === null ? null : reachabilityCriteriaCurrentLink[0],
+        reachability_criteria_dynamic_info: reachabilityCriteriaCurrentLink === null ?
+          undefined :
+          reachabilityCriteriaCurrentLink[1],
+        timestamp: Date.now(),
+        parent: parentStatisticLink,
+      };
+      this.appendReachabilityStatistic(statisticLink, 'pushEvent');
+      const event: ILinkQueueEvent = {
+        type: EventType.Push,
+        link: statisticLink,
+        query: this.query,
+        queueStatistics: {
+          size: this.getSize(),
+          reachabilityRatio: this.reachabilityRatio,
+        },
+      };
+
+      this.materialize(event);
+    }
+    return resp;
+  }
+
+  public override pop(): ILink | undefined {
+    const link = super.pop();
+    if (link !== undefined) {
+      const reachabilityCriteriaCurrentLink = LinkQueueLogger.getReachability(link);
+      const statisticLink: IURLStatistic = {
+        url: link.url,
+        reachability_criteria: reachabilityCriteriaCurrentLink === null ? null : reachabilityCriteriaCurrentLink[0],
+        reachability_criteria_dynamic_info: reachabilityCriteriaCurrentLink === null ?
+          undefined :
+          reachabilityCriteriaCurrentLink[1],
+        timestamp: Date.now(),
+      };
+      this.appendReachabilityStatistic(statisticLink, 'popEvent');
+      const event: ILinkQueueEvent = {
+        type: EventType.Pop,
+        link: statisticLink,
+        query: this.query,
+        queueStatistics: {
+          size: this.getSize(),
+          reachabilityRatio: this.reachabilityRatio,
+        },
+      };
+
+      this.materialize(event);
+    }
+    return link;
+  }
+
+  /**
+   * Materialize the current event into the logger
+   * @param {ILinkQueueEvent} event - Current event
+   */
+  private materialize(event: ILinkQueueEvent): void {
+    const jsonEvent = { ...event, type: EventType[event.type] };
+    this.logger.trace(LinkQueueLogger.LINK_QUEUE_EVENT_NAME, { data: JSON.stringify(jsonEvent) });
+  }
+}
+
+/**
+ * The type of event
+ */
+export enum EventType {
+  Push,
+  Pop,
+}
+
+/**
+ * A link queue event
+ */
+interface ILinkQueueEvent {
+  type: EventType;
+  link: IURLStatistic;
+  query: Algebra.Operation;
+  queueStatistics: IQueueStatistics;
+}
+/**
+ * Statistic of the link queue
+ */
+interface IQueueStatistics {
+  size: number;
+  reachabilityRatio: IReachabilityRatio;
+}
+/**
+ * Ratio of the reachability criteria of the links in the queue
+ */
+interface IReachabilityRatio {
+  pushEvent: Record<string, number>;
```

Comment:
What do string and number mean?

Reply:
It is the criteria with the number of time they have been involve with the event. It is separate because in my other work links can be prune inside of the queue so the number of pushed and popped links will not necessarily match.

Reply:
Can you document it?